### ISSUE TITLE
MediaPipe Selfie Segmentation results are not invoking when moved to other Chrome tabs

### ISSUE THREAD

**Issue by yellagoud** [google/mediapipe#3018]
const videoElement =document.getElementsByClassName('input_video')[0];
selfieSegmentationObj = new SelfieSegmentation({locateFile: (file) => {
    return `https://cdn.jsdelivr.net/npm/@mediapipe/selfie_segmentation/${file}`;
    }});
selfieSegmentationObj.setOptions({
    modelSelection: 1,
    selfieMode: false,
    effect: 'mask',
});

selfieSegmentationObj.onResults(selfieSegmentationResults);

cameraObj = new Camera(videoElement, {
    onFrame: async () => {
          await selfieSegmentationObj.send({image: videoElement});
    },
    width: 1280,
    height: 720
});

cameraObj.start();

function selfieSegmentationResults(results) {
    const canvasElement = document.getElementsByClassName('output_canvas')[0];
    var canvasCtx = canvasElement.getContext('2d');
    canvasCtx.clearRect(0, 0, canvasElement.width, canvasElement.height);
    canvasCtx.save();
    canvasCtx.drawImage(results.segmentationMask, 0, 0, canvasElement.width, canvasElement.height);
}

When we moved from one tab to another tab then callback function "selfieSegmentationResults" is not getting called from the SelfieSegmentation object.

Please, someone, help if there is any other solution that can get the segmented image data even if we moved to another Chrome tab.

**Comment by sgowroji** [comment]
Hi @yellagoud, Can you share any references with error logs or screenshots for the above query. Thanks!

**Comment by yellagoud** [comment]
Hi @sgowroji, 

I didn't find any error logs. After debugging the code I found one issue in the following JS file,
https://cdn.jsdelivr.net/npm/@mediapipe/camera_utils/camera_utils.js

In the above JS, the Camera object is used internally "Window.requestAnimationFrame()" which generally will not execute if the Chrome tab is hidden. The same case is working in the Firefox browser. 

I replaced "Window.requestAnimationFrame()" with setTimeout() then issue is not coming now.

But setTimeout is not a proper solution it may slow down the browser performance. 

Basically, I am doing a background replacement feature that will transmit the replaced background image stream via WebRTC peer connection to others.

so please suggest to me instead of "setTimeout()" what is the best solution???

Thanks.

**Comment by tyrmullen** [comment]
Correct, Camera uses requestAnimationFrame, which browsers may slow down updates to or stop calling back completely when the tab is backgrounded or fully hidden.  This is often a source of confusion (see related Issue #2955), and Chrome definitely has this behavior.  For what you want, setTimeout() is a reasonable approach, provided you call it with good parameters (poll every X ms, instead of trying to poll as fast as possible; don't process every time you get a callback, but make sure you have a new video frame first).

If you're looking for alternative polling methods to try (just for Chrome), then there's also requestVideoFrameCallback and the MediaStreamTrack extensions in the WebCodecs API.  But I don't remember off-hand what the browser backgrounding throttling behavior is for those.  If you try those out, let me know what you find!

**Comment by yellagoud** [comment]
Hi @tyrmullen,

Thank you for responding to my query.  

Somehow video with the background images is getting blurred.  I am trying to draw the complete image onto canvas but the result is image fuzzy. Following is the code i am using,

```
this.canvasCtx.drawImage(results.segmentationMask, 0, 0, this.canvasElement.width, this.canvasElement.height);
this.canvasCtx.globalCompositeOperation = 'source-in';
this.canvasCtx.drawImage(results.image, 0, 0, this.canvasElement.width, this.canvasElement.height);
this.canvasCtx.globalCompositeOperation = 'destination-atop';
this.canvasCtx.drawImage(document.getElementById("connect_bg_image"), 0, 0, this.canvasElement.width, this.canvasElement.height);
```

@this is the image that I am trying to draw https://dev47.beta-wspbx.com/images/connectbackground/office-desktop-1.png but the original size is decreasing onto the canvas.

Could you please help me to solve this issue?

**Comment by tyrmullen** [comment]
In general, we should create a new thread for issues separate from the initially reported one, so if my comment doesn't resolve it, let's move the discussion to a new issue.

Your background image appears to be 640x360 resolution, but the camera resolution for our demos is usually 1280x720.  I'm assuming therefore your canvasElement is 1280x720 to match the camera resolution, which means your background image has to be stretched 2x to fit the canvas, and will therefore look blurry.  You'll need a higher-resolution background image (ideally 1280x720) to fix this.  (I'm assuming the background is what's blurry here?)

**Comment by JohannCarfantan** [comment]
Hello @yellagoud have you fixed this problem? I'm facing the same problem and I would appreciate some help on this 
Thanks

**Comment by yellagoud** [comment]
@JohannCarfantan I downloaded the https://cdn.jsdelivr.net/npm/@mediapipe/camera_utils/camera_utils.js file and replaced following function,
function P(a){ window.requestAnimationFrame(function(){Q(a)}) }
with below code,
function P(a) { setTimeout(function() { Q(a) }, 1000 / 50) }

Moreover, this issue is with only Chrome and doesn't exist this issue in Firefox.
Please let me know if you need any help
Thanks,

**Comment by avicarpio** [comment]
This not working, setTimeout is also throttled by the browser. How can I fix it?

**Comment by avicarpio** [comment]
@yellagoud
